src/mscorlib/src/System/Array.cs

```diff
@@ -1707,13 +1707,14 @@ public static void Reverse<T>(T[] array, int index, int length)
                 ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_InvalidOffLen);
             Contract.EndContractBlock();
 
+            ref T p = ref Unsafe.As<byte, T>(ref array.GetRawSzArrayData());
             int i = index;
             int j = index + length - 1;
             while (i < j)
             {
-                T temp = array[i];
-                array[i] = array[j];
-                array[j] = temp;
+                T temp = Unsafe.Add(ref p, i);
+                Unsafe.Add(ref p, i) = Unsafe.Add(ref p, j);
+                Unsafe.Add(ref p, j) = temp;
```

Comment:
@jkotas, for my edification, what's the explanation for why this is faster?  It's avoiding bounds checks that the JIT is currently unable to elide?

Reply:
It is avoiding bounds checks and covariance checks (for arrays of object references).

Reply:
Got it.

Reply:
`List<T>` is non-covariant so this could be done for `Add` etc?

Reply:
For List, you need to be careful about corrupted state that can lead to buffer overruns or type safety violations. Some of these bugs may be in the user code, but our general standard is that user code race should not turn into type or memory safety violations. I think a better way to deal with the covariance check overhead in general case like List is something like the covariance array wrapper (https://github.com/dotnet/corefx/issues/23689) - the unsafe code is just one line in the covariance array wrapper. The rest is safe.

Reply:
> In theory, the JIT can notice that the source of the value is the same array as what it is assigned to and use that information to ship the covariance check ...

Is skipping the covariance when source and destination are same tracked somewhere?

Reply:
https://github.com/dotnet/coreclr/issues/14574

Reply:
If you load and store to the same array directly the jit already does this optimization (see for instance #14580, which enhanced an existing optimization). But if load and store are disconnected the jit may miss out.

@erozenfeld maybe this is something we could do more generally, once we have built up dataflow?